x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,421 @@
+        processedGlobalCheckpoint = Math.max(processedGlobalCheckpoint, followerLocalCheckpoint);
+        numConcurrentWrites--;
+        assert numConcurrentWrites >= 0;
+        coordinateWrites();
+    }
+
+    private void maybeUpdateMapping(Long minimumRequiredIndexMetadataVersion, Runnable task) {
+        assert Thread.holdsLock(this);
+        if (currentIndexMetadataVersion >= minimumRequiredIndexMetadataVersion) {
+            LOGGER.trace("{} index metadata version [{}] is higher or equal than minimum required index metadata version [{}]",
+                params.getFollowShardId(), currentIndexMetadataVersion, minimumRequiredIndexMetadataVersion);
+            task.run();
+        } else {
+            LOGGER.debug("{} updating mapping, index metadata version [{}] is lower than minimum required index metadata version [{}]",
+                params.getFollowShardId(), currentIndexMetadataVersion, minimumRequiredIndexMetadataVersion);
+            updateMapping(imdVersion -> {
+                retryCounter.set(0);
+                currentIndexMetadataVersion = imdVersion;
+                task.run();
+            });
+        }
+    }
+
+    private void handleFailure(Exception e, Runnable task) {
+        assert e != null;
+        if (shouldRetry(e)) {
+            if (isStopped() == false && retryCounter.incrementAndGet() <= RETRY_LIMIT) {
+                LOGGER.warn("error during follow shard task, retrying...", e);
+                scheduler.accept(RETRY_TIMEOUT, task);
+            } else {
+                markAsFailed(new ElasticsearchException("retrying failed [" + retryCounter.get() +
+                    "] times, aborting...", e));
+            }
+        } else {
+            markAsFailed(e);
+        }
+    }
+
+    private boolean shouldRetry(Exception e) {
+        // TODO: What other exceptions should be retried?
+        return NetworkExceptionHelper.isConnectException(e) ||
+            NetworkExceptionHelper.isCloseConnectionException(e) ||
+            e instanceof ActionTransportException ||
+            e instanceof NodeClosedException ||
+            e instanceof UnavailableShardsException ||
+            e instanceof NoShardAvailableActionException;
+    }
+
+    // These methods are protected for testing purposes:
+    protected void updateMapping(LongConsumer handler) {
+        Index leaderIndex = params.getLeaderShardId().getIndex();
+        Index followIndex = params.getFollowShardId().getIndex();
+
+        ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+        clusterStateRequest.clear();
+        clusterStateRequest.metaData(true);
+        clusterStateRequest.indices(leaderIndex.getName());
+
+        leaderClient.admin().cluster().state(clusterStateRequest, ActionListener.wrap(clusterStateResponse -> {
+            IndexMetaData indexMetaData = clusterStateResponse.getState().metaData().getIndexSafe(leaderIndex);
+            assert indexMetaData.getMappings().size() == 1;
```

Comment:
please put a message in that includes some info that helps to debug?

Reply:
sure